python/ray/experimental/sgd/util.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import ray
+import json
+import logging
+import time
+
+logger = logging.getLogger(__name__)
+
+
+def fetch(oids):
+    if ray.global_state.use_raylet:
+        local_sched_client = ray.worker.global_worker.local_scheduler_client
+        for o in oids:
+            ray_obj_id = ray.ObjectID(o)
+            local_sched_client.reconstruct_objects([ray_obj_id], True)
+    else:
+        for o in oids:
+            plasma_id = ray.pyarrow.plasma.ObjectID(o)
+            ray.worker.global_worker.plasma_client.fetch([plasma_id])
+
+
+def run_timeline(sess, ops, feed_dict={}, write_timeline=False, name=""):
+    if write_timeline:
+        run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
+        run_metadata = tf.RunMetadata()
+        fetches = sess.run(
+            ops,
+            options=run_options,
+            run_metadata=run_metadata,
+            feed_dict=feed_dict)
+        trace = timeline.Timeline(step_stats=run_metadata.step_stats)
+        outf = "timeline-{}-{}.json".format(name, os.getpid())
+        trace_file = open(outf, "w")
+        logger.info("wrote tf timeline to", os.path.abspath(outf))
+        trace_file.write(trace.generate_chrome_trace_format())
+    else:
+        fetches = sess.run(ops, feed_dict=feed_dict)
+    return fetches
+
+
+class Timeline(object):
+    def __init__(self, tid):
+        self.events = []
+        self.offset = 0
+        self.start_time = self.time()
+        self.tid = tid
+
+    def patch_ray(self):
```

Comment:
Why patch the Ray logging? As opposed to just using the existing logging?

Reply:
The ray logging functionality doesn't provide the fine-grained control we have here to capture just one SGD iteration, and also add additional types of events.

Note that this still calls the original log call so you can get overall timelines that way still.

Reply:
Hm, you can just do (assuming xray)

```python
with ray.profile("custom_event"):
    # do one SGD iteration
```

to add the event to the timeline. Is that what you want? Anyway, if there's additional functionality here that's needed for profiling, then in a follow up PR we should just extend the profiling API (so it's available more generally outside of SGD).